File: sdk-csharp/DepHealth.Core/Endpoint.cs
using System.Collections.ObjectModel;
using System.Text.RegularExpressions;

namespace DepHealth;

/// <summary>
/// Dependency endpoint (host + port + labels). Immutable.
/// </summary>
public sealed partial class Endpoint : IEquatable<Endpoint>
{
    private static readonly HashSet<string> ReservedLabels = new(StringComparer.Ordinal)
    {
        "name", "group", "dependency", "type", "host", "port", "critical"
    };

    [GeneratedRegex("^[a-zA-Z_][a-zA-Z0-9_]*$")]
    private static partial Regex LabelNamePattern();

    /// <summary>Endpoint hostname or IP address.</summary>
    public string Host { get; }

    /// <summary>Endpoint port number as string.</summary>
    public string Port { get; }

    /// <summary>Custom Prometheus labels for this endpoint.</summary>
    public IReadOnlyDictionary<string, string> Labels { get; }

    /// <summary>Additional metadata key-value pairs.</summary>
    [Obsolete("Use Labels instead")]
    public IReadOnlyDictionary<string, string> Metadata => Labels;

    /// <summary>Creates an endpoint with host and port.</summary>
    /// <param name="host">Hostname or IP address.</param>
    /// <param name="port">Port number as string.</param>
    public Endpoint(string host, string port)
        : this(host, port, new Dictionary<string, string>())
    {
    }

    /// <summary>Creates an endpoint with host, port, and custom labels.</summary>
    /// <param name="host">Hostname or IP address.</param>
    /// <param name="port">Port number as string.</param>
    /// <param name="labels">Custom Prometheus labels.</param>
    public Endpoint(string host, string port, IDictionary<string, string> labels)
    {
        Host = host ?? throw new ArgumentNullException(nameof(host));
        Port = port ?? throw new ArgumentNullException(nameof(port));
        Labels = new ReadOnlyDictionary<string, string>(
            new Dictionary<string, string>(labels ?? new Dictionary<string, string>()));
    }

    /// <summary>Returns the port as an integer.</summary>
    public int PortAsInt() => int.Parse(Port, System.Globalization.CultureInfo.InvariantCulture);

    /// <summary>
    /// Validates that the label name is valid: [a-zA-Z_][a-zA-Z0-9_]* and not reserved.
    /// </summary>
    public static void ValidateLabelName(string name)
    {
        if (string.IsNullOrEmpty(name))
        {
            throw new ValidationException("label name must not be empty");
        }

        if (!LabelNamePattern().IsMatch(name))
        {
            throw new ValidationException(
                $"label name must match [a-zA-Z_][a-zA-Z0-9_]*, got '{name}'");
        }

        if (ReservedLabels.Contains(name))
        {
            throw new ValidationException(
                $"label name '{name}' is reserved");
        }
    }

    /// <summary>
    /// Validates all keys in the labels dictionary.
    /// </summary>
    public static void ValidateLabels(IReadOnlyDictionary<string, string> labels)
    {
        foreach (var key in labels.Keys)
        {
            ValidateLabelName(key);
        }
    }

    /// <inheritdoc />
    public bool Equals(Endpoint? other)
    {
        if (other is null)
        {
            return false;
        }

        return Host == other.Host && Port == other.Port;
    }

    /// <inheritdoc />
    public override bool Equals(object? obj) => Equals(obj as Endpoint);

    /// <inheritdoc />
    public override int GetHashCode() => HashCode.Combine(Host, Port);

    /// <inheritdoc />
    public override string ToString() => $"{Host}:{Port}";
}

File: sdk-java/dephealth-core/src/main/java/biz/kryukov/dev/dephealth/Endpoint.java
package biz.kryukov.dev.dephealth;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Dependency endpoint (host + port + labels). Immutable.
 */
public final class Endpoint {

    /** Label name pattern: starts with a letter or _, followed by letters, digits, _. */
    private static final Pattern LABEL_NAME_PATTERN =
            Pattern.compile("^[a-zA-Z_][a-zA-Z0-9_]*$");

    /** Labels reserved by the SDK — cannot be overridden via WithLabel/label(). */
    public static final Set<String> RESERVED_LABELS = Set.of(
            "name", "group", "dependency", "type", "host", "port", "critical"
    );

    private final String host;
    private final String port;
    private final Map<String, String> metadata;

    public Endpoint(String host, String port) {
        this(host, port, Map.of());
    }

    public Endpoint(String host, String port, Map<String, String> metadata) {
        this.host = Objects.requireNonNull(host, "host");
        this.port = Objects.requireNonNull(port, "port");
        this.metadata = metadata == null ? Map.of() : Collections.unmodifiableMap(metadata);
    }

    public String host() {
        return host;
    }

    public String port() {
        return port;
    }

    public int portAsInt() {
        return Integer.parseInt(port);
    }

    /**
     * Returns custom labels of the endpoint.
     *
     * @return unmodifiable map of labels
     */
    public Map<String, String> labels() {
        return metadata;
    }

    /**
     * @deprecated Use {@link #labels()} instead.
     */
    @Deprecated
    public Map<String, String> metadata() {
        return metadata;
    }

    /**
     * Validates the label name against the Prometheus naming pattern.
     *
     * @param name label name
     * @throws ValidationException if the name does not match the pattern or is reserved
     */
    public static void validateLabelName(String name) {
        Objects.requireNonNull(name, "label name");
        if (!LABEL_NAME_PATTERN.matcher(name).matches()) {
            throw new ValidationException(
                    "label name must match " + LABEL_NAME_PATTERN.pattern()
                            + ", got '" + name + "'");
        }
        if (RESERVED_LABELS.contains(name)) {
            throw new ValidationException(
                    "label name '" + name + "' is reserved and cannot be used as a custom label");
        }
    }

    /**
     * Validates all labels in the map: names and reserved status.
     *
     * @param labels label map
     * @throws ValidationException if any name is invalid
     */
    public static void validateLabels(Map<String, String> labels) {
        if (labels == null) {
            return;
        }
        for (String key : labels.keySet()) {
            validateLabelName(key);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Endpoint ep)) {
            return false;
        }
        return host.equals(ep.host) && port.equals(ep.port);
    }

    @Override
    public int hashCode() {
        return Objects.hash(host, port);
    }

    @Override
    public String toString() {
        return host + ":" + port;
    }
}
